lib/ex_unit/lib/ex_unit/doc_test.ex

```diff
@@ -234,30 +234,50 @@ defmodule ExUnit.DocTest do
     do_import = Keyword.get(opts, :import, false)
 
     extract(module)
-    |> filter_by_opts(opts)
+    |> filter_by_opts(module, opts)
     |> Stream.with_index()
     |> Enum.map(fn {test, acc} ->
       compile_test(test, module, do_import, acc + 1)
     end)
   end
 
-  defp filter_by_opts(tests, opts) do
+  defp filter_by_opts(tests, module, opts) do
     except = Keyword.get(opts, :except, [])
 
     case Keyword.fetch(opts, :only) do
       {:ok, []} ->
         []
 
       {:ok, only} ->
-        tests
-        |> Stream.reject(&(&1.fun_arity in except))
-        |> Stream.filter(&(&1.fun_arity in only))
+        filter_tests(module, tests, except, only)
 
       :error ->
         Stream.reject(tests, &(&1.fun_arity in except))
     end
   end
 
+  defp filter_tests(module, tests, except, only) do
+    {filtered_tests, fun_arities} =
+      for test <- tests,
+          test.fun_arity not in except,
+          test.fun_arity in only,
+          reduce: {[], []} do
+        {tests, fun_arities} -> {[test | tests], [test.fun_arity | fun_arities]}
+      end
+
+    case only -- fun_arities do
+      [] ->
+        filtered_tests
+
+      [_ | _] = undefined_fun_arities ->
+        functions = Enum.map_join(undefined_fun_arities, ", ", fn {f, a} -> "#{f}/#{a}" end)
+
+        raise Error,
+          module: module,
+          message: "undefined or private function(s): #{functions} in module #{inspect(module)}"
```

Comment:
```suggestion
        functions = Enum.map_join(undefined_fun_arities, ", ", fn {f, a} -> Exception.format_mfa(module, f, a) end)

        raise Error,
          module: module,
          message: "undefined or private function(s) given to doctest: #{functions}"
```

Reply:
I think this makes it less readable when there are multiple functions undefined. They all belong to the same module. 

May be, at least, list them each on new line padded with 4 spaces, like:
```
** (ExUnit.DocTest.Error) <file_name>: undefined or private function(s) given to doctest:

    Module.foo/0
    Module.bar/1
    
<stacktrace>
```

Reply:
Multi line looks great. Your call!